### PR TITLE
ApiError parsing located to common place

### PR DESCRIPTION
**Pull request by WestDiscGolf**
FitbitExceptions re-jigged

Resubmitting PR after manually making changes to updated project structure

### CODE REVIEW

File: Fitbit.Portable.Tests/AccessTokenTests.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        {
+            string content = SampleDataHelper.GetContent("ApiError-Request-StaleToken.json");
+
+            try
+            {
+                OAuth2Helper.ParseAccessTokenResponse(content);
+            }
+            catch (FitbitTokenException exception)
+            {
+                // can only use ShouldThrow on async funcs of Task
```
Comment (mxa0079) [inline Fitbit.Portable.Tests/AccessTokenTests.cs]: At some point we need to move all tests to async. Not for RC... higher priorities.

File: Fitbit.Portable/OAuth2/OAuth2Helper.cs [hunk 2]
```diff
@@ -1,65 +1,34 @@
...
         internal static string Base64Encode(string plainText)
         {
-            var plainTextBytes = System.Text.Encoding.UTF8.GetBytes(plainText);
-            return System.Convert.ToBase64String(plainTextBytes);
+            var plainTextBytes = Encoding.UTF8.GetBytes(plainText);
+            return Convert.ToBase64String(plainTextBytes);
         }
     }
-}
+}
```
Comment (mxa0079) [inline Fitbit.Portable/OAuth2/OAuth2Helper.cs]: @WestDiscGolf Outstanding job at cleaning up this class!

File: Fitbit.Portable/OAuth2/OAuth2AutoRefreshInterceptor.cs [hunk 3]
```diff
@@ -42,7 +42,7 @@ public Task<HttpResponseMessage> InterceptRequest(HttpRequestMessage request, Ca
                     }
                     else if (response.Result.RequestMessage.Headers.Contains(CUSTOM_HEADER))
                     {
-                        throw new FitbitException("We received an unexpected stale token response -- during the retry for a call whose token we just refreshed", response.Result.StatusCode);
+                        throw new FitbitTokenException(message: $"We received an unexpected stale token response - during the retry for a call whose token we just refreshed {response.Result.StatusCode}");
```
Comment (mxa0079) [inline Fitbit.Portable/OAuth2/OAuth2AutoRefreshInterceptor.cs]: Nice improvement!

File: Fitbit.Portable/OAuth2/OAuth2AutoRefreshInterceptor.cs [hunk 4]
```diff
@@ -53,17 +53,8 @@ public Task<HttpResponseMessage> InterceptRequest(HttpRequestMessage request, Ca
...
-            foreach (JToken error in errors)
-            {
-                var apiError = JsonConvert.DeserializeObject<ApiError>(error.ToString());
-                if (apiError.ErrorType == "expired_token")
-                    return true;
-            }
-
-            return false;
+            var errors = new JsonDotNetSerializer().Errors(responseBody);
+            return errors.Any(error => error.ErrorType == "expired_token");
```
Comment (mxa0079) [inline Fitbit.Portable/OAuth2/OAuth2AutoRefreshInterceptor.cs]: Much cleaner!

File: Fitbit.Portable/JsonDotNetSerializerExtensions.cs [hunk 5]
```diff
@@ -9,6 +9,23 @@ namespace Fitbit.Api.Portable
     internal static class JsonDotNetSerializerExtensions
     {
         /// <summary>
+        /// Parses the error structure which is common when errors are raised from the api
+        /// </summary>
+        /// <param name="serializer"></param>
+        /// <param name="errorJson"></param>
+        /// <returns></returns>
+        internal static List<ApiError> Errors(this JsonDotNetSerializer serializer, string errorJson)
```
Comment (mxa0079) [inline Fitbit.Portable/JsonDotNetSerializerExtensions.cs]: @WestDiscGolf I think this should be named `ParseFitbitErrors` or something similar. Can you amend the PR?

**Comment by mxa0079** [comment]
@WestDiscGolf Overall, outstanding job and centralizing logic and removing duplicate code. Great refactoring!

File: Fitbit.Portable.Tests/AccessTokenTests.cs (hunk 1 above)
Comment (WestDiscGolf) [inline Fitbit.Portable.Tests/AccessTokenTests.cs]: The helper wasn't so I couldn't use the fluent assertions exception testing. Agree will improve after RC.

File: Fitbit.Portable/OAuth2/OAuth2Helper.cs (hunk 2 above)
Comment (WestDiscGolf) [inline Fitbit.Portable/OAuth2/OAuth2Helper.cs]: @mxa0079 Thanks :-)

File: Fitbit.Portable/JsonDotNetSerializerExtensions.cs (hunk 5 above)
Comment (WestDiscGolf) [inline Fitbit.Portable/JsonDotNetSerializerExtensions.cs]: Done :-)

**Comment by WestDiscGolf** [comment]
@mxa0079 Thanks, means a lot :-)

**Pull request merged**